### PR TITLE
Fly around in the lobby (if the server mods allow it)

### PR DESCRIPTION
**Pull request by johnshaughnessy**
Adds the ability to fly around in the lobby.
~TODO: Add "enter in vr" button to watch from immersive vr mode.~
~TODO: Only allow this option if the server moderator enables it.~
TODO: All in-world menus are currently turned off. We should enable volume controls, "open link" and inspect/focus.
TODO: Make sure I didn't break scene.html (How do I test this locally?)

### CODE REVIEW

File: src/systems/scene-preview-camera-system.js [hunk 1]
```diff
@@ -14,10 +17,18 @@ export class ScenePreviewCameraSystem {
   }
 
   tick() {
+    viewingCamera = viewingCamera || document.getElementById("viewing-camera");
+    uiRoot = uiRoot || document.getElementById("ui-root");
+    const isGhost = uiRoot && uiRoot.firstChild && uiRoot.firstChild.classList.contains("isGhost");
+    const entered = viewingCamera && viewingCamera.sceneEl.is("entered");
     for (const el of this.entities) {
```
Comment (robertlong) [inline src/systems/scene-preview-camera-system.js]: This should probably be a regular for loop. `for of` allocates.

File: src/components/scene-preview-camera.js [hunk 2]
```diff
@@ -54,6 +55,10 @@ AFRAME.registerComponent("scene-preview-camera", {
   tick2: function() {
     const hubsSystems = this.el.sceneEl.systems["hubs-systems"];
 
+    uiRoot = uiRoot || document.getElementById("ui-root");
+    const isGhost = uiRoot && uiRoot.firstChild && uiRoot.firstChild.classList.contains("isGhost");
```
Comment (robertlong) [inline src/components/scene-preview-camera.js]: There's a lot of checking for this "isGhost" class. I'm a little worried about the perf of the number of calls to `contains` every frame.

**Comment by gfodor** [comment]
fyi, there's already a facility for the "inspect, volume controls, etc"
menus icymi - we do that if the user does not have the relevant
permissions, so we might want to treat this lobby mode as a non-privileged
session basically

File: src/systems/scene-preview-camera-system.js (hunk 1 above)
Comment (johnshaughnessy) [reply src/systems/scene-preview-camera-system.js]: fixed

File: src/components/scene-preview-camera.js (hunk 2 above)
Comment (johnshaughnessy) [reply src/components/scene-preview-camera.js]: will only check for `isGhost` if the scene is not entered yet

**Comment by misslivirose** [comment]
This is just crossing my radar with the addition of the What's New tag since I was out when this was discussed - if we don't already, we need to put a warning on rooms where this is enabled. I've already had people express discomfort with this idea and IMO even on properties that we're not maintaining, we should have UI that indicates this mode.

**Comment by misslivirose** [comment]
I'll think about a design where that can be made clear, but I think this a pattern where we should err on the side of being overly cautious about the ways that someone could be burned by not knowing there's an invisible avatar listening over their shoulder.

**Comment by misslivirose** [comment]
Some questions: 

1. If I enter a room where this feature is disabled, and an administrator changes it while I'm in the room, what happens? Is there any way that I am informed that the "ghost" feature is now active?

2. How does this interact with future decisions that we make about fall-off and audio attenuation? Many people use the visual metaphor within a room to understand who in the group is present. This feature changes the dynamic of those spaces significantly to a greater extent, because people now drop the visual metaphor of knowing the range of their audio reach. 

Some thoughts: 

From what I recall (and since #350 is still open), we still don't have a mechanic for VR users to even understand who is in the lobby. That feels like something that should be surfaced in VR _especially_ given that there is now a feature where some deployments of Hubs are now going to allow lobby viewers to move freely and break the spatial metaphor. 

I suspect that most people aren't going to realize that the fact that the additional button on this page means that others will be able to freely move about with no indicators to them in the room. Given the sensitivity of again, breaking that metaphor that people who are near you will be made known, this should be more explicitly communicated when this feature is enabled. 

This feels especially important given that HMC _doesn't_ have this enabled by default, and people may be bringing the mental model of proximity / general lobby viewing from one instance to another.

Personally, I'd feel very uncomfortable using Hubs instances that had this feature enabled, given that I wouldn't have any visual way of knowing if someone was invisibly in the lobby up close listening to me and recording me. Without this setting, doing that is a lot harder but that's still a privacy vector to consider, IMO.

**Comment by misslivirose** [comment]
Not a complete design, but at a minimum I'd like to see a subtitle on the 'Enter Room' button in this design that says something explicit like 'This room has enabled lobby ghosts' or maybe even more directly, 'Other participants may be invisible and inaudible'.

**Comment by johnshaughnessy** [comment]
Liv brings up some good points. We might want to pull the lobby ghosts feature from hubs clouds instances until we have a better story around surfacing it to users so that they know what to expect. I think it's OK for other instances to have their own rules / expectations but we can do better by most users with good defaults.

Adding a few related concerns here for when we revisit this topic: 
- Traversing the room from the lobby is possible via the object list
- The lobby camera view is not visible from within the room
- The avatar volume controls allows easy listening at a distance.
- A modded client can access any of the audio streams and play them at any volume
- A modded client can move the camera to anywhere in the room

I support us making it easy to be a good actor and harder to be a bad actor, as long as we don't accidentally give people a false sense of security in the process. 

Related: "Disappearing messages", "preventing screenshots", and DRM are implemented differently in different apps. It's important that people recognize that while these make it slightly more inconvenient for a bad actor to break the social contract, there are no guarantees here.

**Comment by gfodor** [comment]
In terms of the users understanding this capability, isn’t the extra button
on the landing page the “notification” of the feature, in addition to the
actual mechanism to use it?

I agree that awareness of presence is a big problem in VR. We have the
count, but the lack of a UI toolkit is basically the root cause of this and
other poor UX in immersive mode around information not being displayed.

**Comment by gfodor** [comment]
If the current button is not serving as a sufficient notice (I don’t have
the copy handy) perhaps we can try to adjust that accordingly so it solves
both problems?

**Comment by gfodor** [comment]
(Both problems being “users have informed consent” and “there is a way to
enter lobby fly mode” - the user list problem I have no good solution for)

**Comment by misslivirose** [comment]
> In terms of the users understanding this capability, isn’t the extra button
on the landing page the “notification” of the feature, in addition to the
actual mechanism to use it?

I think that the issue with this is that the action button here is designed for the people taking the action of moving in the lobby; it isn't the flow that the user looking for the "Enter Room" is going to take, so the likelihood of them noticing that there's an extra button and reading the fine print on there feels high to me.

**Comment by mohoman** [comment]
From our perspective, we do not really need the concept of a "Lobby" but more of "Spectators"
What is most important for us is to get more people in the room simply to listen to a speaker or watch a video. So in that case, the "Spectator does not need to be able to speak or have an avatar. But we would want other avatars to be able to see them, even if it's just a 2D ghost icon or something.
Things the "Spectator" could do would be: move around the room, chat with people, listen to conversations, watch videos, and sit in chairs.
If the lobby feature could be optimized to maximize the number of "Spectators" in a room, that is what we would like to have.

**Pull request merged**